omniscidb/QueryBuilder/QueryBuilder.cpp

```diff
@@ -0,0 +1,3072 @@
+  auto name = name_.empty() ? "approx_count_dist" : name_ + "_approx_count_dist";
+  return {builder_, agg, name, true};
+}
+
+BuilderExpr BuilderExpr::approxQuantile(double val) const {
+  if (!expr_->type()->isNumber()) {
+    throw InvalidQueryError() << "Unsupported type for sum aggregate: "
+                              << expr_->type()->toString();
+  }
+  if (val < 0.0 || val > 1.0) {
+    throw InvalidQueryError()
+        << "ApproxQuantile expects argument between 0.0 and 1.0 but got " << val;
+  }
+  Datum d;
+  d.doubleval = val;
+  auto cst = makeExpr<Constant>(builder_->ctx_.fp64(), false, d);
+  auto agg = makeExpr<AggExpr>(
+      builder_->ctx_.fp64(), AggType::kApproxQuantile, expr_, false, cst);
+  auto name = name_.empty() ? "approx_quantile" : name_ + "_approx_quantile";
+  return {builder_, agg, name, true};
+}
+
+BuilderExpr BuilderExpr::sample() const {
+  auto agg = makeExpr<AggExpr>(expr_->type(), AggType::kSample, expr_, false, nullptr);
+  auto name = name_.empty() ? "sample" : name_ + "_sample";
+  return {builder_, agg, name, true};
+}
+
+BuilderExpr BuilderExpr::singleValue() const {
+  if (expr_->type()->isVarLen()) {
+    throw InvalidQueryError() << "Varlen type " << expr_->type()->toString()
+                              << " is not suported for single value aggregate.";
+  }
+  auto agg =
+      makeExpr<AggExpr>(expr_->type(), AggType::kSingleValue, expr_, false, nullptr);
+  auto name = name_.empty() ? "single_value" : name_ + "_single_value";
+  return {builder_, agg, name, true};
+}
+
+BuilderExpr BuilderExpr::agg(const std::string& agg_str, double val) const {
+  static const std::unordered_map<std::string, AggType> agg_names = {
+      {"count", AggType::kCount},
+      {"count_dist", AggType::kCount},
+      {"count_distinct", AggType::kCount},
+      {"count dist", AggType::kCount},
+      {"count distinct", AggType::kCount},
+      {"sum", AggType::kSum},
+      {"min", AggType::kMin},
+      {"max", AggType::kMax},
+      {"avg", AggType::kAvg},
+      {"approx_count_dist", AggType::kApproxCountDistinct},
+      {"approx_count_distinct", AggType::kApproxCountDistinct},
+      {"approx count dist", AggType::kApproxCountDistinct},
+      {"approx count distinct", AggType::kApproxCountDistinct},
+      {"approx_quantile", AggType::kApproxQuantile},
+      {"approx quantile", AggType::kApproxQuantile},
+      {"sample", AggType::kSample},
+      {"single_value", AggType::kSingleValue},
+      {"single value", AggType::kSingleValue}};
+  static const std::unordered_set<std::string> distinct_names = {
+      "count_dist", "count_distinct", "count dist", "count distinct"};
+  auto agg_str_lower = boost::algorithm::to_lower_copy(agg_str);
+  if (!agg_names.count(agg_str_lower)) {
+    throw InvalidQueryError() << "Unknown aggregate name: " << agg_str;
+  }
+
+  auto kind = agg_names.at(agg_str_lower);
+  if (kind == AggType::kApproxQuantile && val == HUGE_VAL) {
```

Comment:
What value is huge? 3? :)

Reply:
`HUGE_VAL` is a constant used as `+INF` double value. In this function, we use this value by default to allow users to skip `val` arg for most of the aggregates. And if `APPROX QUANTILE` aggregate is used but `val` is missing, then an exception is thrown.